### ISSUE TITLE
Can't enter normal mode (or any other than insert) at all

### ISSUE THREAD

**Issue by mrjones2014** [godbout/kindaVim.theapp#66]
Updated to KindaVim 18 this morning and now it doesn't seem to be working at all. Regardless of whether I have the normal mode setting to escape or a sequence like `jk` I cannot enter normal mode in any app regardless of what family it's in.

**Comment by mrjones2014** [comment]
Yup, double checked. I removed the accessibility permission and re added it. Also tried restarting kindaVim. Will try rebooting my computer later today, I've got some zoom meetings to attend first.

**Comment by godbout** [comment]
alright, thanks.
if it still doesn't work, can you 1) send the file located at `~/Library/Preferences/mo.com.sleeplessmind.kindaVim.plist` to guill@sleeplessmind.com.mo 2) switch on the Viming window / icon change and tell me if anything seems to be happening.

**Comment by mrjones2014** [comment]
Interesting... I toggled this setting off and it started working, then toggled it back on and it still works. Must have just been some weird edge case or something. 

![CleanShot-2022-04-25-at-12 19 13](https://user-images.githubusercontent.com/8648891/165131179-8d9e8538-c83b-4b95-8df3-e13e93edbc8b.png)

**Comment by godbout** [comment]
and also glad it works again :D

**Comment by mrjones2014** [comment]
Actually, the issue is back 🤔

**Comment by godbout** [comment]
> if it still doesn't work, can you 1) send the file located at `~/Library/Preferences/mo.com.sleeplessmind.kindaVim.plist` to [guill@sleeplessmind.com.mo](mailto:guill@sleeplessmind.com.mo) 2) switch on the Viming window / icon change and tell me if anything seems to be happening.

send me your plist then by email, so that i load kV with your settings and see.

toggling the Characters Window setting doesn't anymore?

also:

> any other app you have installed that could steal escape or jk? Hammerspoon?

**Comment by godbout** [comment]
just in case, can you try with several apps also? just to make sure this is not an Electron thing again. (just sending things in the air. goal is to pinpoint asap so that you can use kV again without issue.)

**Comment by mrjones2014** [comment]
Seems to be intermittent because it's working again this morning, without me doing anything other than locking my computer overnight, and unlocking it this morning. I've emailed you the plist file. I have hammerspoon installed, but I don't bind anything to escape or `jk`.

Here's my whole hammerspoon config: https://github.com/mrjones2014/dotfiles/tree/master/.config/hammerspoon

**Comment by mrjones2014** [comment]
Hmm. It must be something else on my system because now I'm having the same issue with kv17. :(

**Comment by godbout** [comment]
i've checked your `plist`. looked fine. use kV with it and no issue here.

but sorry i'm getting  a bit confused now. is it working back or not then? in any case the intermittence makes it even more bizarre. from what you've described and your answers, it would seem that kV's event tap doesn't receive the keystrokes. but if doesn't have AX access you'll get the prompt. so the only thing is can think of (out of kV) is that something is getting the keys and not passing them back, or on a wrong tap level.

**Comment by godbout** [comment]
> Hmm. It must be something else on my system because now I'm having the same issue with kv17. :(

ah. not cool.
anything you've installed lately? changed?
i can send you kV19 alpha, should fail the same i guess, but i can add logs so that we can pinpoint what's happening on your system. would you mind trying? sorry about that. no other report so far so gonna require a bit of work from your side too to debug 😬️

**Comment by mrjones2014** [comment]
I'd be more than happy to try kv19 beta. I'm looking through my command history to see if any new homebrew packages might be causing issue. kindavim doesn't need Input Monitoring permission right? it doesn't request it as far as I can tell.

**Comment by godbout** [comment]
also, to confirm it's not the `plist`, can you back it up, rename it, and start with kV again and see if it changes anything?

**Comment by godbout** [comment]
> I'd be more than happy to try kv19 beta. I'm looking through my command history to see if any new homebrew packages might be causing issue. kindavim doesn't need Input Monitoring permission right? it doesn't request it as far as I can tell.

no, just AX permissions. that includes event taps. although now that you mention, you're on macOS 12.3.1? (not a beta?) if it's on a beta, you never know. Apple restricts stuff more and more. i'm developing on the latest stable (12.3.1).

**Comment by godbout** [comment]
give me an hour or two and i'll upload the 19a here. thanks.

**Comment by mrjones2014** [comment]
Renamed the plist to plist.tmp and still have the same issue.

>you're on macOS 12.3.1? (not a beta?)

correct, I'm on latest stable.

**Comment by godbout** [comment]
> FWIW restarting seems to fix the issue for a while, then after a while it comes up again. I'll take note of anything out of the ordinary that happens in between this time, and how long it takes.

you mean rebooting macOS right? or restarting kV?

**Comment by mrjones2014** [comment]
Rebooting macOS.

**Comment by mrjones2014** [comment]
It's just annoying because macOS has this bug where every time I restart, all my display settings get reset to defaults, and it decides to use the monitor that I have rotated 90 degrees into portrait mode as my default display 😂

**Comment by godbout** [comment]
> It's just annoying because macOS has this bug where every time I restart, all my display settings get reset to defaults, and it decides to use the monitor that I have rotated 90 degrees into portrait mode as my default display 😂

i will not take responsibility for that one 😅️

(i almost never reboot here, so i guess if it was just kV, i'd see the issue too. installing your hs setup just to make sure. will add the log then and push the update but not much experience with logging for the Console.app so gonna take an hour probably.)

**Comment by godbout** [comment]
[kV19a.zip](https://github.com/godbout/kindaVim.theapp/files/8564435/kV19a.zip)

not notarized (coz sometimes it takes ages) so you will need to 1) explicitly allow it in the Security Preferences thing 2) probably remove and add back kV from the Accessibility permissions.

after that you should see messages in the `Console.app`. screenshot coming below:

**Comment by godbout** [comment]
1. choose your computer
2. show info option (remove debug, else a lot of noise)
3. search for kindaVim

![Screen Shot 2022-04-26 at 23 06 18](https://user-images.githubusercontent.com/121373/165332291-3a209114-121d-4761-b85a-5fc53ec8a8f8.png)

you should start seeing stuff. the interesting part will be to see what it prints when it stops working.

**Comment by godbout** [comment]
forgot, you need to press `Start streaming` to record the logs.

**Comment by mrjones2014** [comment]
seems to just keep asking for accessibility permissions even when it has them already 🤔 

FWIW I haven't been able to reproduce the issue since I rebooted macOS again a few hours ago though

**Comment by godbout** [comment]
> seems to just keep asking for accessibility permissions even when it has them already 🤔

that's because you're dealing with two apps with the same bundle ID but different signatures. you have to remove the current entry of kindaVim from the list (with the `-` button), and add it again by pressing the `Grant Accessibility blah blah` button from the beta.

> FWIW I haven't been able to reproduce the issue since I rebooted macOS again a few hours ago though

ok, fingers crossed, but if this just needed a reboot, you'll then deserve the:

**Comment by godbout** [comment]
![tumblr_mwa55k0ZsD1sgl0ajo1_500](https://user-images.githubusercontent.com/121373/165351647-c47baab6-b20d-46b8-9995-acec28ae5c8f.gif)

**Comment by mrjones2014** [comment]
Lol, I tried rebooting yesterday too though

**Comment by godbout** [comment]
well if you rebooted yesterday and it stopped working today then the issue will probably appear again. which version you're running now? 17? 19a? if it happens again please run 19a and let's check the logs. i guess it's something with your system but still, kV deserves to run properly!

**Comment by mrjones2014** [comment]
Haven't been able to reproduce. I think we can close this.

**Comment by mrjones2014** [comment]
(Running kv18 btw)

**Comment by godbout** [comment]
> Haven't been able to reproduce. I think we can close this.

good to hear. but still a bit bizarre.

**Comment by mrjones2014** [comment]
Ran into this again. Rebooting seems to fix the issue like 80% of the time. Occasionally need to reboot twice.

**Comment by mrjones2014** [comment]
Once I can reproduce again I'll gather the logs as you've shown above.

**Comment by godbout** [comment]
i think you'll get the medal for the weirdest issue reported.

don't use kV19a, you'll lose the Characters Window goodies and stuff introduced in kV19. i'll add some logs again and build a kV20a soon.

**Comment by mrjones2014** [comment]
I'm using regular kv19, installed from the updater, not the zip you sent 👍

**Comment by godbout** [comment]
yes. i'm mentioning this because kV19 (from the updater) does not have any logging functionality. kV19a was a one-off. haven't studied logging enough to know if macOS cleans automatically or if i need to do something proper. will be the same for kV20a for the time being. down along the way i'll integrate some logging functionality into kV properly but not the right time yet.

**Comment by mrjones2014** [comment]
Wanna hear something really strange?

It started working again this morning, randomly, literally without me doing anything. I was keeping it in its non-working state so that I could get relevant logs once you sent the updated build with additional logging -- but now its randomly working again 😅  wtf

**Comment by godbout** [comment]
wait. it really bothers me. i'm gonna do it now.

**Comment by godbout** [comment]
another thing. i see you have Kitty and CleanShotX in the `Off Family`. when kV stops working, can you remove both those apps and see if you're still having the issue? this is a long shot too. but i'm having issues with macOS sometimes when i plug an external screen and my MacBook is in clamshell mode. when i switch space, some background app is stealing the focus (in my case iTerm, or Yoink). if it's iTerm, because it's in my `Off Family`, kV will not enter Normal Mode even if it looks like i'm on another app. it's because the focus is actually on iTerm, but iTerm is not visible. as Kitty seems to be a terminal app, i don't know, just wondering. going out on a limb here.

**Comment by godbout** [comment]
[kV20a.zip](https://github.com/godbout/kindaVim.theapp/files/8632129/kV20a.zip)

please follow the step described somewhere above in a big screenshot.
you'll probably need first to:
1. remove kV from Accessibility, start kV20a that will ask you to add, add and you'll be ok. **but you need to remove the current entry first**.
2. accept to open an app unidentified blah blah blah from the same Security Preferences 
3. press the "Start streaming" and filter by "kindaVim" to see kV's messages

**Comment by mrjones2014** [comment]
>- are you working on 1Password and Secure Fields?
> - when kV stops working, do you know if hammerspoon can register your keystrokes? if you don't know give it a try when kV stops working, to see if it only affects kV
> - do you live with a roommate you like to do bad pranks?

When it stops working, it doesn't work _anywhere_. When it stops working, Hammerspoon is still able to get my keystrokes (I have a keyboard shortcut for locking and sleeping, they both work). No roommates lol 

I'll install kv20a and start the logs, though I can't reproduce the issue now 😅  it seems I just have to wait until it randomly stops working. I have no idea what triggers it.

**Comment by mrjones2014** [comment]
I'll just leave the log streaming running until I notice it stop working, and take note of roughly what time I noticed it not working.

**Comment by godbout** [comment]
> When it stops working, it doesn't work _anywhere_. 
yeah i got that. but maybe when developing or debugging on 1P, there's some stuff running with macOS. no idea. anyways, as hm is working, shouldn't be the issue.

> When it stops working, Hammerspoon is still able to get my keystrokes (I have a keyboard shortcut for locking and sleeping, they both work). No roommates lol
what a pity. if hm wasn't working i could have blamed it on them ☹️

**Comment by godbout** [comment]
> I'll just leave the log streaming running until I notice it stop working, and take note of roughly what time I noticed it not working.

honestly that seems more painful that necessary. you mentioned before that restarting kV doesn't help. i put logs in the setup of the app too. so what i would do if i were you is just use kV, and when it stops working, then you quit, start the logging, and start kV. it should log the setting up, and we'll be able to see where it stops, and hopefully get to what's going on!!!

**Comment by godbout** [comment]
👀️

**Comment by mrjones2014** [comment]
Still hasn't come up again lol

**Comment by godbout** [comment]
ok let's not talk about it then.

**Comment by godbout** [comment]
👀️👀️

**Comment by mrjones2014** [comment]
I ran into it one other time, but it fixed itself without me doing anything about an hour later. Haven't been able to reproduce with the latest version.

**Comment by godbout** [comment]
you can't see it (i hope?) but i'm currently shaking my head from left to right and my eyes are a little bit like this -> 🙄️

ok let's keep this open. please update when necessary 😭️

**Comment by godbout** [comment]
i was able to reproduce. although it sounds like i'm taking credit for this. what i did was 1) watch WWDC 2) go to bed 3) using my Mac in the morning. i indeed couldn't get kV to receive key events from macOS. mouse events were still working (i catch mouse events so that if you click somewhere you're leaving Normal Mode and going back to macOS). it's pretty bizarre as the tap is not disabled and still works, as i receive the mouse events, but nothing will do for the keyboard. after reboot, everything is back to normal.

what i'll do now:
1. check around if this is some kind of known issue
2. post on StackOverflow (and get downvoted to hell)
3. report a Feedback to Apple
4. wait for it to happen again and see if it's just kV or if other apps are affected as well

**Comment by godbout** [comment]
i know stuff that breaking as soon as Apple announces new stuff but still. i don't know. you. me. i would expect this to show up on other people's computers too. not cool!

will update once i know more. (probably this thread will live forever.)

**Comment by godbout** [comment]
btw @mrjones2014, can i ping you again about your whole feel about kV? you were so enthusiastic on your first day 😅️
some frustrations? some stuff you would have liked to be done differently? thanks in advance.

**Comment by godbout** [comment]
it's been a while.

**Comment by godbout** [comment]
i knew the thoughts and prayers thing was working somehow.

**Comment by mrjones2014** [comment]
lol guess what.... you're gonna hate me....

**Comment by mrjones2014** [comment]
it just happened again 😬

**Comment by godbout** [comment]
two hours after i ask LOL

fuck. well i guess a reboot every three months is acceptable? 👀️
never happened again in my case. that is a very weird issue. macOS would stop sending kV key presses while the event tap was still working.

**Comment by mrjones2014** [comment]
wait idk how it took me this long to realize, but I actually didn't run into this issue -- I ran into kindaVim just crashing. I just realized it wasn't in my menubar. I relaunched it and am not having this issue.

So I guess there's a separate issue causing a crash.

**Comment by godbout** [comment]
> wait idk how it took me this long to realize,

💔️
+
![Unknown](https://user-images.githubusercontent.com/121373/181278086-956cb4e9-8730-444c-8937-45ee187e03c6.jpeg)

**Comment by godbout** [comment]
> So I guess there's a separate issue causing a crash.

do you remember which apps you were using before you didn't realize kV was not on? (🙄️)

i've fixed some holes recently when i found that some Electron apps were returning NEGATIVE caret location. there's still some places in kV that _assume_ the apps are not too dumb, but i need to tighten this up yeah. personally i have 0 crash when i use kV (like always), but it happened once or twice when i curate some weird apps for The Wizard.

will investigate. thank you! (and no i still don't have your... you know.)

**Comment by godbout** [comment]
strange. Slack should be in the `Key Mapping`, so no text calculation. shouldn't crash. except of course if you were on Slack while having ph running in the background.

**Comment by godbout** [comment]
happened once again a few weeks ago. checked other apps (like Alfred) that i know have to use the event tap stream to do something, and they don't receive the events either. so it's a macOS bug. reported: https://github.com/feedback-assistant/reports/issues/390

**Comment by godbout** [comment]
a bit more info on that:

1. this happens when another app has locked the a SecureInput. like they locked and forgot to release or something
2. some tips about how to find which app is currently locking a SecureInput: https://apple.stackexchange.com/questions/331557/is-there-a-way-to-fix-or-disable-secure-input
3. a bit more info here: https://forum.keyboardmaestro.com/t/how-can-i-disable-secure-input-when-km-is-stuck/9361/9
4. and here: https://wiki.keyboardmaestro.com/assistance/Secure_Input_Problem
5. and here: https://smile.helpscoutdocs.com/article/402-textexpander-and-secure-input
6. and also here: https://iterm2-discuss.narkive.com/qd8JsW9w/cannot-turn-off-secure-keyboard-entry

it's finally been made obvious to me because, as point 6 says, iTerm2 makes it obvious when a SecureInput is currently ON. then the other points listed above help you find the culprit, and kill the app.
